### PR TITLE
Transferring website to use drasil generators

### LINKED ISSUE #2524: Swift; missing doxygen generation
Right now, if you click on any of the "Generated Code Documentation" links on the main website page, it will 404.

caused by:
image

image

### PR DESCRIPTION
**Pull request by Ant13731**
This looks like a lot of changed lines, but most of it is generated stuff/the drasil picture. A first pass at the new website. It's not really perfect and the filepaths get tossed around to multiple files which makes everything a little messy.
I decided to separate the sections into different files to copy the drasil-example style and to make things a little less messy.

Closes #2524 

Should there be something in the gitignore file for drasil-website/Website?

Edit: I'm not sure what to change to get the website to be the main webpage for Drasil.

Some notes: Fixing #1686 will break the website, but its an easy fix from there (just make everything lowercase)

### CODE REVIEW

File: code/drasil-website/stack.yaml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+# resolver: nightly-2015-09-21
+# resolver: ghc-7.10.2
+#
+# The location of a snapshot can be provided as a file or url. Stack assumes
+# a snapshot provided as a file might change, whereas a url resource does not.
+#
+# resolver: ./custom-snapshot.yaml
+# resolver: https://example.com/snapshots/2018-01-01.yaml
+resolver:
+  url: https://raw.githubusercontent.com/commercialhaskell/stackage-snapshots/master/lts/17/15.yaml
```
Comment (balacij) [inline code/drasil-website/stack.yaml:21]: Just a very minor thing, the version we use for the rest of the packages is 16.31, does downgrading cause any issues? Also, it's good to know that many of the drasil-* packages are compatible with 17.15.

File: code/drasil-website/package.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    main: Drasil.Website.Main
+    source-dirs: ./
+    other-modules:
+    - Drasil.Website.Website
+    - Drasil.Website.Analysis
+    - Drasil.Website.CaseStudy
+    - Drasil.Website.Documentation
+    - Drasil.Website.Example
+    - Drasil.Website.Graphs
+    - Drasil.Website.Introduction
```
Comment (balacij) [inline code/drasil-website/package.yaml:48]: Was `hpack` not automatically picking these up?

File: code/drasil-website/Website/Drasil_Website.html [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+        <h1>Generated Examples</h1>
+        <p class="paragraph">
+          Each of the case studies contain their own generated PDF and HTML reports, and in some cases, their own generated code.
+        </p>
+        <ul class="list">
+          <li>
+            DblPendulum 
+            <ul class="list">
+              <li>
+                DblPendulumSRS <a href="/home/ant13731/Drasil/code/deploy/examples/DblPendulum/srs/DblPendulum_SRS.html">[HTML]</a> <a href="/home/ant13731/Drasil/code/deploy/examples/DblPendulum/srs/DblPendulum_SRS.pdf">[PDF]</a>
```
Comment (balacij) [inline code/drasil-website/Website/Drasil_Website.html:207]: Links will break here. I think you can just use relative hrefs instead of absolute hrefs.

**Comment by balacij** [comment]
> I'm not sure what to change to get the website to be the main webpage for Drasil.

As long as the website folder is copied and pasted into the `deploy/` folder after you manually ran the CI locally, you'll be fine. The `drasil-bot` is setup to grab that whole folder and paste it into the `gh-pages` branch, which is what GitHub Pages uses to serve content.

If you want to test that, you can fork the repo, create a bot account and set it up as per https://github.com/JacquesCarette/Drasil/pull/2461, merge this branch into your fork's `master` branch and then you'll be able to see your bot run through the steps too. Alternatively, I can give you collaborator access to my fork and you can do the same steps without needing to create a bot account if you don't want to.

File: code/drasil-website/package.yaml (hunk 2 above)
Comment (Ant13731) [reply code/drasil-website/package.yaml:48]: Yea, it wouldn't compile unless I put the extra module names in the package.yaml file.

**Comment by Ant13731** [comment]
I made a second account but I'm not too sure how to get it to run. I set up all the repo secrets on my main account as in #2461. Is the personal access token from the bot account? And will merging this branch to master (on my forked repo) activate the website automatically?

**Comment by Ant13731** [comment]
I've got this branch merged on the forked repo. Do I just have to wait for the github actions to finish? (https://github.com/Ant13731/Drasil)

**Comment by balacij** [comment]
Yes, the personal access token should be created on behalf of the bot account. When any commit is made to master (PR counts as a commit too by the way). Yes, merging this branch to master on your forked repo will activate the website build automatically -- as seen here: https://github.com/Ant13731/Drasil/runs/2932290286

**Comment by balacij** [comment]
The last step, _Deploy_, will be the final thing that pulls in your `gh-pages` branch, updates it, and pushes it to your own github pages. Then, you'll be able to access the website at: https://ant13731.github.io/Drasil/

**Comment by Ant13731** [comment]
Okay, sweet! Thanks!

**Comment by balacij** [comment]
Hmm, that's weird. My `gh-pages` branch is usually behind this repos + divergent. Did you pull this branch into your fork and then merge it into your remote's master branch?

**Comment by balacij** [comment]
Ah, looking at your deploy step, it looks like it found no work to commit https://github.com/Ant13731/Drasil/runs/2932570396?check_suite_focus=true

**Comment by Ant13731** [comment]
I used GitHub desktop to transfer the `transferringWebsite` branch to my local forked repo and then pushed from there. Could that be it?

**Comment by Ant13731** [comment]
Oh, should I just add another commit then?

**Comment by balacij** [comment]
The GitHub desktop transfer is fine, your commits were there; https://github.com/JacquesCarette/Drasil/compare/master...Ant13731:master

You might however need to try running through the CI steps locally to see what is actually placed into the deploy folder and where. It might just be that the scripts are failing to copy the new website over the old files.

**Comment by balacij** [comment]
Oh, wait, looking again, it looks like your build was actually caught by the build check when it built the website. https://github.com/Ant13731/Drasil/runs/2932570396?check_suite_focus=true#step:19:8

Sorry, I completely forgot about that!

I think https://github.com/Ant13731/Drasil/pull/1 will fix it (thanks GitHub online editor!).

**Comment by Ant13731** [comment]
I think one of my tokens may have been wrong, so I'm going to try and generate a new personal access token and try again.

**Comment by balacij** [comment]
Looks like it's up :smile:  https://ant13731.github.io/Drasil/

**Comment by Ant13731** [comment]
I got it to work! Now I think I should see if the old website folder could/should be deleted? Also a lot of the links are broken, but it's there!

File: code/drasil-website/package.yaml (hunk 2 above)
Comment (balacij) [reply code/drasil-website/package.yaml:48]: Oh, did you only have 1 or 2 listed in the package.yaml file? Does it still happen if you entirely remove the `other-modules` list section?

Comment (Ant13731) [reply code/drasil-website/package.yaml:48]: Oh, yea I only had Drasil.Website.Website there before. I'll try removing them all and see if it does it automatically.

**Comment by balacij** [comment]
Nice! I'd say so, we can always retrieve it from the git history if need be. The links I think are still related to the issue I mentioned above -- using absolute references still I think.

Since we now also directly rely on the drasil-* packages for the website, it might be good to also add it to the `code/stack.yaml` under the packages list; https://github.com/JacquesCarette/Drasil/blob/56fa1ada9a7798247073c6e05ee55bd9453ac718/code/stack.yaml#L39-L52 so that it gets compiled during the tests as well -- another safety measure to make sure that the website will actually build for the full Build & Deploy (.github/workflows/Build.yaml) instead of just the mini-build (.github/workflows/Test.yaml). If we do this, we should remove the `drasil-website/Makefile` and merge it's contents with the `code/Makefile` so that we can also share the `.stack-work`.

**Comment by Ant13731** [comment]
I tried adding code/drasil-website/Website to the .gitignore file but the index file still shows up when using git add *.

I'm also having a bit of trouble transferring the makefile commands.

**Comment by balacij** [comment]
You probably also need to remove it from the tracked files -- `git rm -r --cached code/drasil-website/Website`.

Anything in particular that I could help with for the Makefile? We can keep it in the drasil-website folder if it's too much of a hassle.

**Comment by Ant13731** [comment]
I'm not really sure how to get all the needed environment variables to go into the stack build command (and some of the variables have the same name but a different filepath)

**Comment by balacij** [comment]
Just so you know, I think that having it in the main Makefile will simplify the scripts, so some things might need to just get removed because they'll end up redundant. 

I _think_ they should, so as long as we add `drasil-website` to the `code/stack.yaml` file.

**Comment by balacij** [comment]
A good start is to start with very basic new targets for the main Makefile (no environment variables yet) that just compile the `drasil-website` executable via `stack install`. Once you have it building, it should be erroring out if you try to edit it, then you can try using the existing deploy script to use the installed binary (note that the `website` binary will then be "installed" into the `.local/bin`, so it should become accessible from anywhere via the terminal).

**Comment by Ant13731** [comment]
Okay, I see. So once stack install is called, I can use stack build and stack exec in the script?

**Comment by balacij** [comment]
Well, `stack install` is secretly a `stack build` with a `--copy-bins`, and I think you should be able to do it before as well considering we do the same method for the examples too.

Ah, thanks for the PR, I just saw that while I posting this.

Okay, so, here's what I would do:

After the `stack install website` in the `code/Makefile`, I would build the website right there using the variables from there (it's usually just passed through other areas, but in this, they will exist for sure, maybe under different names however). Afterwards, I would alter the `deploy_stage.sh`'s `build_website` function to be more like the `copy_*` functions whereby it just copies the contents of the website build artifacts. The `deploy_stage.sh` file will no longer need to run `stack build website` & `stack exec website` since we already build it via `make website` in the main `code/Makefile`.

Aside: when you pass environment variables to a program via bash, it doesn't quite work like how it would in an imperative language. You need to use the following form in one single line `variable=... variable=... mainBinaryName <optionally arguments for the actual binary being run>`. In your file, I think you were trying to pass variables using multiple variable definitions just before running the command.

**Comment by Ant13731** [comment]
Okay, I see. I think its starting to work. So right now I have a `make website` target that gives the environment variables to the stack exec command. If the test & build pass then I will probably close this PR and open a new PR with the forked repo. Thanks for all your help!

**Comment by Ant13731** [comment]
The other branch's build tests worked, so I'm going to close this and open a new PR.

**Pull request closed without merging**